### ISSUE TITLE
How to set http proxy while calling rest endpoint

### ISSUE THREAD

**Issue by anshul-kumar-agl** [paperbits/paperbits-common#7]
Hello, when calling a rest endpoint as below, I need to mention http proxy so that request goes through given proxy. How can we do same?

import { HttpClient, HttpRequest, HttpResponse, HttpMethod, HttpHeader } from "@paperbits/common/http";

response = await this.HttpClient.send<T>(httpRequest);

**Comment by azaslonov** [comment]
Hi @anshul-kumar-agl, when you say "mention http proxy", do you mean some kind of header? If possible, please provide more details or analogy of what are you referring to.

**Comment by anshul-kumar-agl** [comment]
> Hi @anshul-kumar-agl, when you say "mention http proxy", do you mean some kind of header? If possible, please provide more details or analogy of what are you referring to.

Like we mention a proxy url to go every request through in 'request' package as below.

const request = require('request');

const options = {
  url: '....',
  method:'GET',
  proxy:'http://localhost:9000',
  headers: {
    'Authorization': '....'
  }
};

We have to route each request through a http proxy url like above.

**Comment by azaslonov** [comment]
Sorry, I'm still confused. Probably you want to have some kind of base URL, so that your request would look like:
```
const options = {
    url: '/my-relative-url',
    method:'GET',
    proxy:'http://localhost:9000',
};
```
Is that right?

If not, how would actual HTTP request look like if, let's say, you used something like Postman?

**Comment by anshul-kumar-agl** [comment]
> Sorry, I'm still confused. Probably you want to have some kind of base URL, so that your request would look like:
> 
> ```
> const options = {
>     url: '/my-relative-url',
>     method:'GET',
>     proxy:'http://localhost:9000',
> };
> ```
> 
> Is that right?
> 
> If not, how would actual HTTP request look like if, let's say, you used something like Postman?

Url would be complete url. Just that I want request to route through proxy url which internally routes to organisation network. The code I shared using package 'request' because the url mentioned is reached by routing through http proxy mentioned in proxy property.

Can we set options like above to HttpClient or HttpRequest of your package.

**Comment by azaslonov** [comment]
Basically, our client is just a wrapper over XmlHttpRequest, which doesn't seem to have options like `proxy` (https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)

This is why I'm asking for an analogy. So, if you did raw HTTP request through Postman or Fiddler how would it look like?

**Comment by anshul-kumar-agl** [comment]
> Basically, our client is just a wrapper over XmlHttpRequest, which doesn't seem to have options like `proxy` (https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)
> 
> This is why I'm asking for an analogy. So, if you did raw HTTP request through Postman or Fiddler how would it look like?

The request we do through postman has url and authorisation header. I don't see mentioning anything else. One difference is, since we have application on local machine which sets http proxy on localhost:9000, so application requests, web requests and presumably postman request route through http proxy and hence we get back data.

**Comment by azaslonov** [comment]
Right, postman also cannot do that because it uses the same XmlHttpRequest.

So, the only option, I think, is to have a setting that, depending on the environment, would switch between your localhost:9000 and production endpoint.

**Comment by joshua-mng** [comment]
@anshul-kumar-agl I think best way to achieve what you are trying to do is to create new custom HttpClient based on code from `@paperbits-common/http/httpClient`, where you can implement proxy like behaviour yourself.
